File: frontend/packages/contracts/src/rateLimit.ts
import { z } from "zod";
import { IdSchema, UtcDateTimeSchema } from "./common";

export const RateLimitEndpointSchema = z.enum(["quote", "historical", "orders", "other"]);
export type RateLimitEndpoint = z.infer<typeof RateLimitEndpointSchema>;

export const RateLimitWindowSchema = z.enum(["second", "minute", "day"]);
export type RateLimitWindow = z.infer<typeof RateLimitWindowSchema>;

/**
 * One limit window (D27). `brokerLimit` is set by the broker and read-only; `novaLimit` is our own
 * safety limit (≤ broker). `used` is the peak for second/minute windows (rolling) and the count so
 * far for the day window, which resets at `resetsAt`.
 */
export const RateLimitRuleSchema = z
  .strictObject({
    window: RateLimitWindowSchema,
    brokerLimit: z.number().int().positive(),
    novaLimit: z.number().int().positive(),
    used: z.number().int().min(0),
    resetsAt: UtcDateTimeSchema.nullable(),
  })
  .refine((r) => r.novaLimit <= r.brokerLimit, {
    message: "novaLimit must be less than or equal to brokerLimit",
    path: ["novaLimit"],
  })
  .refine((r) => r.used <= r.brokerLimit, {
    message: "used must be less than or equal to brokerLimit",
    path: ["used"],
  })
  .refine((r) => (r.window === "day") === (r.resetsAt !== null), {
    message: "resetsAt is set only for the day window",
    path: ["resetsAt"],
  });
export type RateLimitRule = z.infer<typeof RateLimitRuleSchema>;

export const RateLimitSchema = z
  .strictObject({
    accountId: IdSchema,
    endpoint: RateLimitEndpointSchema,
    rules: z.array(RateLimitRuleSchema).min(1),
    throttledToday: z.number().int().min(0),
    updatedAt: UtcDateTimeSchema,
  })
  .refine((l) => new Set(l.rules.map((r) => r.window)).size === l.rules.length, {
    message: "rules must have unique windows",
    path: ["rules"],
  });
export type RateLimit = z.infer<typeof RateLimitSchema>;

/** Request body for `PATCH /broker/rate-limits/{accountId}/{endpoint}`: changes one NOVA limit. */
export const RateLimitUpdateSchema = z.strictObject({
  window: RateLimitWindowSchema,
  novaLimit: z.number().int().positive(),
});
export type RateLimitUpdate = z.infer<typeof RateLimitUpdateSchema>;

File: backend/libs/nova_contracts/src/nova_contracts/rate_limit.py
"""Rate limits (D27): mirror `frontend/packages/contracts/src/rateLimit.ts`."""

from typing import Annotated, Literal, Self

from pydantic import Field, model_validator

from nova_contracts.common import Contract, Id, UtcDateTime

RateLimitEndpoint = Literal["quote", "historical", "orders", "other"]
RateLimitWindow = Literal["second", "minute", "day"]
Positive = Annotated[int, Field(gt=0)]


class RateLimitRule(Contract):
    window: RateLimitWindow
    broker_limit: Positive
    nova_limit: Positive
    used: Annotated[int, Field(ge=0)]
    resets_at: UtcDateTime | None

    @model_validator(mode="after")
    def _limits(self) -> Self:
        if self.nova_limit > self.broker_limit:
            raise ValueError("novaLimit must be less than or equal to brokerLimit")
        if self.used > self.broker_limit:
            raise ValueError("used must be less than or equal to brokerLimit")
        if (self.window == "day") != (self.resets_at is not None):
            raise ValueError("resetsAt is set only for the day window")
        return self


class RateLimit(Contract):
    account_id: Id
    endpoint: RateLimitEndpoint
    rules: Annotated[list[RateLimitRule], Field(min_length=1)]
    throttled_today: Annotated[int, Field(ge=0)]
    updated_at: UtcDateTime

    @model_validator(mode="after")
    def _unique_windows(self) -> Self:
        if len({rule.window for rule in self.rules}) != len(self.rules):
            raise ValueError("rules must have unique windows")
        return self


class RateLimitUpdate(Contract):
    """Body of `PATCH /broker/rate-limits/{accountId}/{endpoint}`: changes one NOVA limit."""

    window: RateLimitWindow
    nova_limit: Positive
